### ISSUE TITLE
Accessing enum by enum value in switch errors with: TS2342: An index expression argument must be of type 'string', 'number', 'symbol', or 'any'.

### ISSUE THREAD

**Issue by vladeck** [Microsoft/TypeScript#13116]
Typescript 2.0.7

Code:

```typescript
enum EFoo {
    Foo1,
    Foo2,
    Foo3,
}

class Foo {
    public bar(f: EFoo): void {
        switch (f) {
            case EFoo.Foo1:
                // ...
            break;

            case EFoo.Foo2:
                // ...
            break;

            case EFoo.Foo3:
                // ...
            break;

            default:
                throw new Error(`Foo::bar: Unknown EFoo '${EFoo[f]}'.`);
        }
    }
}
```

Errors with (using command line tsc):

`error TS2342: An index expression argument must be of type 'string', 'number', 'symbol', or 'any'.`

Errors with (within VSCode 1.8.1 with bundled tsc 2.1.4):

`message: 'Type 'never' cannot be used as an index type.'`

However, this code does not produce an error:

```typescript
class Foo {
    public bar(f: EFoo): void {
        const fType: string = EFoo[f];

        switch (f) {
            case EFoo.Foo1:
                // ...
            break;

            case EFoo.Foo2:
                // ...
            break;

            case EFoo.Foo3:
                // ...
            break;

            default:
                throw new Error(`Foo::bar: Unknown EFoo '${fType}'.`);
        }
    }
}
```

Since the more detailed documentation on enums is lacking, I'm not sure if this is an issue with the Typescript or my understanding of it. I truly apologize if it the latter.

**Comment by mhegazy** [comment]
In the first example the compiler has concluded that all possible values of this enum has been exhausted by the preceding switch labels, and that is why the type of `f` is narrowed to `never`.

in the second example you have used `f` before its type was narrowed, as `f` can still be `EFoo.Foo1 | EFoo.Foo2 | EFoo.Foo3`

**Comment by vladeck** [comment]
I understand, but does it have to result in an error? What if I add another enum entry and want to make sure that that new entry is handled by my code (that is why there is a default)?

**Comment by mhegazy** [comment]
> I understand, but does it have to result in an error? 

`f` being narrowed to `never` is not an error. using `f` to index into an enum is; since the compiler does not know what type it has and can not do any changes. consider casting to `any`.

**Comment by RyanCavanaugh** [comment]
See also http://stackoverflow.com/questions/39419170/how-do-i-check-that-a-switch-block-is-exhaustive-in-typescript
